src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/Kafka/KafkaProduceSyncDeliveryHandlerIntegration.cs

```diff
@@ -0,0 +1,179 @@
+                if (span is null)
+                {
+                    Logger.Warning("Unexpected null span for Kafka Producer with delivery handler");
+                    return CallTargetState.GetDefault();
+                }
+
+                var newAction = CachedWrapperDelegate<TActionOfDeliveryReport>.CreateWrapper(handler, span);
+
+                Action<ITypedDeliveryHandlerShimAction> updateHandlerAction = inst => inst.Handler = newAction;
+
+                // store the call to update the handler variable as state
+                // so we update it at the _end_ of the constructor
+                return new CallTargetState(scope: null, state: updateHandlerAction);
+            }
+            catch (Exception ex)
+            {
+                Logger.Error(ex, "Error creating wrapped delegate for delivery report");
+                return CallTargetState.GetDefault();
+            }
+        }
+
+        /// <summary>
+        /// OnMethodEnd callback
+        /// </summary>
+        /// <typeparam name="TTarget">Type of the target</typeparam>
+        /// <param name="instance">Instance value, aka `this` of the instrumented method.</param>
+        /// <param name="exception">Exception instance in case the original code threw an exception.</param>
+        /// <param name="state">Calltarget state value</param>
+        /// <returns>A response value, in an async scenario will be T of Task of T</returns>
+        public static CallTargetReturn OnMethodEnd<TTarget>(TTarget instance, Exception exception, CallTargetState state)
+        {
+            if (state.State is Action<ITypedDeliveryHandlerShimAction> updateHandlerAction
+             && instance.TryDuckCast<ITypedDeliveryHandlerShimAction>(out var shim))
+            {
+                try
+                {
+                    updateHandlerAction.Invoke(shim);
+                }
+                catch (Exception ex)
+                {
+                    Logger.Error(ex, "There was an error updating the delivery report handler to ");
+                    // Not ideal to close the span here immediately, but as we can't trace the result,
+                    // we don't really have a choice
+                    state.Scope?.Span.Finish();
+                }
+            }
+
+            return CallTargetReturn.GetDefault();
+        }
+
+        /// <summary>
+        /// Helper method used by <see cref="CachedWrapperDelegate{TActionDelegate}"/> to create a delegate
+        /// </summary>
+        /// <param name="originalHandler">The original delivery report handler </param>
+        /// <param name="span">A <see cref="Span"/> that can be manipulated when the action is invoked</param>
+        /// <typeparam name="TDeliveryReport">Type of the delivery report</typeparam>
+        /// <returns>The wrapped action</returns>
+        public static Action<TDeliveryReport> WrapAction<TDeliveryReport>(Action<TDeliveryReport> originalHandler, Span span)
```

Comment:
Do you think there would be issues if we did not pass a `Span` argument and had the returned delegate just call `Tracer.Instance.ActiveScope`?

Reply:
Yeah, I think that could be problematic... The callback could be executed at any point, and I don't think the Thread is even guaranteed. so `Tracer.Instance.ActiveScope.Span` is very likely to _not_ be the `Span` we pass in here I think...

Reply:
Understood, thanks for clarifying